File: linkis-computation-governance/linkis-manager/linkis-application-manager/src/main/scala/org/apache/linkis/manager/rm/external/yarn/YarnResourceRequester.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.linkis.manager.rm.external.yarn

import org.apache.linkis.common.utils.{Logging, Utils}
import org.apache.linkis.manager.common.conf.RMConfiguration
import org.apache.linkis.manager.common.entity.resource.{
  CommonNodeResource,
  NodeResource,
  ResourceType,
  YarnResource
}
import org.apache.linkis.manager.common.errorcode.ManagerCommonErrorCodeSummary._
import org.apache.linkis.manager.common.exception.{RMErrorException, RMWarnException}
import org.apache.linkis.manager.rm.external.domain.{
  ExternalAppInfo,
  ExternalResourceIdentifier,
  ExternalResourceProvider
}
import org.apache.linkis.manager.rm.external.request.ExternalResourceRequester
import org.apache.linkis.manager.rm.utils.RequestKerberosUrlUtils

import org.apache.commons.lang3.StringUtils
import org.apache.http.{HttpHeaders, HttpResponse}
import org.apache.http.client.methods.HttpGet
import org.apache.http.impl.client.HttpClients
import org.apache.http.util.EntityUtils

import java.text.MessageFormat
import java.util
import java.util.Base64
import java.util.concurrent.ConcurrentHashMap

import scala.collection.JavaConverters._
import scala.collection.mutable.ArrayBuffer

import org.json4s.JsonAST._
import org.json4s.JValue
import org.json4s.jackson.JsonMethods.parse

class YarnResourceRequester extends ExternalResourceRequester with Logging {

  private val HASTATE_ACTIVE = "active"

  private val rmAddressMap: util.Map[String, String] = new ConcurrentHashMap[String, String]()

  private def getAuthorizationStr(provider: ExternalResourceProvider) = {
    val user = provider.getConfigMap.getOrDefault("user", "").asInstanceOf[String]
    val pwd = provider.getConfigMap.getOrDefault("pwd", "").asInstanceOf[String]
    val authKey = user + ":" + pwd
    Base64.getMimeEncoder.encodeToString(authKey.getBytes)
  }

  override def requestResourceInfo(
      identifier: ExternalResourceIdentifier,
      provider: ExternalResourceProvider
  ): NodeResource = {
    val rmWebAddress = getAndUpdateActiveRmWebAddress(provider)
    logger.info(s"rmWebAddress: $rmWebAddress")
    val queueName = identifier.asInstanceOf[YarnResourceIdentifier].getQueueName

    def getYarnResource(jValue: Option[JValue]) = jValue.map(r =>
      new YarnResource(
        (r \ "memory").asInstanceOf[JInt].values.toLong * 1024L * 1024L,
        (r \ "vCores").asInstanceOf[JInt].values.toInt,
        0,
        queueName
      )
    )

    def maxEffectiveHandle(queueValue: Option[JValue]): Option[YarnResource] = {
      val metrics = getResponseByUrl("metrics", rmWebAddress, provider)
      val totalResouceInfoResponse = (
        (metrics \ "clusterMetrics" \ "totalMB").asInstanceOf[JInt].values.toLong,
        (metrics \ "clusterMetrics" \ "totalVirtualCores").asInstanceOf[JInt].values.toLong
      )
      queueValue.map(r => {
        val absoluteCapacity = r \ "absoluteCapacity" match {
          case jDecimal: JDecimal =>
            jDecimal.values.toDouble
          case jDouble: JDouble =>
            jDouble.values
          case _ =>
            0d
        }
        val effectiveResource = absoluteCapacity
        new YarnResource(
          math
            .floor(effectiveResource * totalResouceInfoResponse._1 * 1024L * 1024L / 100)
            .toLong,
          math.floor(effectiveResource * totalResouceInfoResponse._2 / 100).toInt,
          0,
          queueName
        )
      })
    }

    var realQueueName = "root." + queueName

    def getQueue(queues: JValue): Option[JValue] = queues match {
      case JArray(queue) =>
        queue.foreach { q =>
          val yarnQueueName = (q \ "queueName").asInstanceOf[JString].values
          if (yarnQueueName == realQueueName) return Some(q)
          else if (realQueueName.startsWith(yarnQueueName + ".")) {
            return getQueue(getChildQueues(q))
          }
        }
        None
      case JObject(queue) =>
        if (
            queue
              .find(_._1 == "queueName")
              .exists(_._2.asInstanceOf[JString].values == realQueueName)
        ) {
          Some(queues)
        } else {
          val childQueues = queue.find(_._1 == "childQueues")
          if (childQueues.isEmpty) None
          else getQueue(childQueues.map(_._2).get)
        }
      case _ => None
    }

    def getChildQueues(resp: JValue): JValue = {
      val queues = resp \ "childQueues" \ "queue"

      if (
          queues != null && queues != JNull && queues != JNothing && null != queues.children && queues.children.nonEmpty
      ) {
        logger.info(s"queues:$queues")
        queues
      } else resp \ "childQueues"
    }

    def getQueueOfCapacity(queues: JValue): Option[JValue] = queues match {
      case JArray(queue) =>
        queue.foreach { q =>
          val yarnQueueName = (q \ "queueName").asInstanceOf[JString].values
          if (yarnQueueName == realQueueName) return Some(q)
          else if ((q \ "queues").toOption.nonEmpty) {
            val matchQueue = getQueueOfCapacity(getChildQueuesOfCapacity(q))
            if (matchQueue.nonEmpty) return matchQueue
          }
        }
        None
      case JObject(queue) =>
        if (
            queue
              .find(_._1 == "queueName")
              .exists(_._2.asInstanceOf[JString].values == realQueueName)
        ) {
          return Some(queues)
        } else if ((queues \ "queues").toOption.nonEmpty) {
          val matchQueue = getQueueOfCapacity(getChildQueuesOfCapacity(queues))
          if (matchQueue.nonEmpty) return matchQueue
        }
        None
      case _ => None
    }

    def getChildQueuesOfCapacity(resp: JValue) = resp \ "queues" \ "queue"

    def getResources() = {
      val resp = getResponseByUrl("scheduler", rmWebAddress, provider)
      val schedulerType =
        (resp \ "scheduler" \ "schedulerInfo" \ "type").asInstanceOf[JString].values
      if ("capacityScheduler".equals(schedulerType)) {
        realQueueName = queueName
        val childQueues = getChildQueuesOfCapacity(resp \ "scheduler" \ "schedulerInfo")
        val queue = getQueueOfCapacity(childQueues)
        val queueOption = Option(queue) match {
          case Some(queue) => queue
          case None =>
            logger.debug(s"cannot find any information about queue $queueName, response: " + resp)
            throw new RMWarnException(
              YARN_NOT_EXISTS_QUEUE.getErrorCode,
              MessageFormat.format(YARN_NOT_EXISTS_QUEUE.getErrorDesc, queueName)
            )
        }
        val queueInfo = queueOption.get.asInstanceOf[JObject]
        (
          maxEffectiveHandle(queue).get,
          getYarnResource(queue.map(_ \ "resourcesUsed")).get,
          (queueInfo \ "maxApps").asInstanceOf[JInt].values.toInt,
          (queueInfo \ "numPendingApps").asInstanceOf[JInt].values.toInt,
          (queueInfo \ "numActiveApps").asInstanceOf[JInt].values.toInt
        )
      } else if ("fairScheduler".equals(schedulerType)) {
        if ("root".equals(queueName)) {
          // get cluster total resource
          val queue = (resp \ "scheduler" \ "schedulerInfo" \ "rootQueue")
          val rootQueue: Option[JValue] = Some(queue)
          val rootQueueInfo = rootQueue.get.asInstanceOf[JObject]
          (
            getYarnResource(rootQueue.map(_ \ "maxResources")).get,
            getYarnResource(rootQueue.map(_ \ "usedResources")).get,
            (rootQueueInfo \ "maxApps").asInstanceOf[JInt].values.toInt,
            0,
            0
          )
        } else {
          val childQueues = getChildQueues(resp \ "scheduler" \ "schedulerInfo" \ "rootQueue")
          val queue = getQueue(childQueues)
          if (queue.isEmpty || queue.get == null) {
            logger.debug(s"cannot find any information about queue $queueName, response: " + resp)
            throw new RMWarnException(
              YARN_NOT_EXISTS_QUEUE.getErrorCode,
              MessageFormat.format(YARN_NOT_EXISTS_QUEUE.getErrorDesc, queueName)
            )
          }
          val queueInfo = queue.get.asInstanceOf[JObject]
          (
            getYarnResource(queue.map(_ \ "maxResources")).get,
            getYarnResource(queue.map(_ \ "usedResources")).get,
            (queueInfo \ "maxApps").asInstanceOf[JInt].values.toInt,
            (queueInfo \ "numPendingApps").asInstanceOf[JInt].values.toInt,
            (queueInfo \ "numActiveApps").asInstanceOf[JInt].values.toInt
          )
        }
      } else {
        logger.debug(
          s"only support fairScheduler or capacityScheduler, schedulerType: $schedulerType , response: " + resp
        )
        throw new RMWarnException(
          ONLY_SUPPORT_FAIRORCAPA.getErrorCode,
          MessageFormat.format(ONLY_SUPPORT_FAIRORCAPA.getErrorDesc(), schedulerType)
        )
      }
    }

    Utils.tryCatch {
      val yarnResource = getResources()
      val nodeResource = new CommonNodeResource
      nodeResource.setMaxResource(yarnResource._1)
      nodeResource.setUsedResource(yarnResource._2)
      nodeResource.setMaxApps(yarnResource._3)
      nodeResource.setNumPendingApps(yarnResource._4)
      nodeResource.setNumActiveApps(yarnResource._5)
      nodeResource
    }(t => {
      throw new RMErrorException(
        YARN_QUEUE_EXCEPTION.getErrorCode,
        YARN_QUEUE_EXCEPTION.getErrorDesc,
        t
      )
    })
  }

  override def requestAppInfo(
      identifier: ExternalResourceIdentifier,
      provider: ExternalResourceProvider
  ): java.util.List[ExternalAppInfo] = {

    val rmWebAddress = getAndUpdateActiveRmWebAddress(provider)

    val queueName = identifier.asInstanceOf[YarnResourceIdentifier].getQueueName

    def getYarnResource(jValue: Option[JValue]) = jValue.map(r =>
      new YarnResource(
        (r \ "allocatedMB").asInstanceOf[JInt].values.toLong * 1024L * 1024L,
        (r \ "allocatedVCores").asInstanceOf[JInt].values.toInt,
        0,
        queueName
      )
    )

    val realQueueName = "root." + queueName

    def getAppInfos(): Array[ExternalAppInfo] = {
      val resp = getResponseByUrl("apps", rmWebAddress, provider)
      resp \ "apps" \ "app" match {
        case JArray(apps) =>
          val appInfoBuffer = new ArrayBuffer[YarnAppInfo]()
          apps.foreach { app =>
            val yarnQueueName = (app \ "queue").asInstanceOf[JString].values
            val state = (app \ "state").asInstanceOf[JString].values
            if (yarnQueueName == realQueueName && (state == "RUNNING" || state == "ACCEPTED")) {
              val appInfo = new YarnAppInfo(
                (app \ "id").asInstanceOf[JString].values,
                (app \ "user").asInstanceOf[JString].values,
                state,
                (app \ "applicationType").asInstanceOf[JString].values,
                getYarnResource(Some(app)).get
              )
              appInfoBuffer.append(appInfo)
            }
          }
          appInfoBuffer.toArray
        case _ => new ArrayBuffer[YarnAppInfo](0).toArray
      }
    }

    Utils.tryCatch(getAppInfos().toList.asJava)(t => {
      throw new RMErrorException(
        YARN_APPLICATION_EXCEPTION.getErrorCode,
        YARN_APPLICATION_EXCEPTION.getErrorDesc,
        t
      )
    })
  }

  override def getResourceType: ResourceType = ResourceType.Yarn

  private def getResponseByUrl(
      url: String,
      rmWebAddress: String,
      provider: ExternalResourceProvider
  ) = {
    val httpGet = new HttpGet(rmWebAddress + "/ws/v1/cluster/" + url)
    httpGet.addHeader("Accept", "application/json")
    val authorEnable: Any = provider.getConfigMap.get("authorEnable");
    var httpResponse: HttpResponse = null
    authorEnable match {
      case flag: Boolean =>
        if (flag) {
          httpGet.addHeader(HttpHeaders.AUTHORIZATION, "Basic " + getAuthorizationStr(provider))
        }
      case _ =>
    }
    val kerberosEnable: Any = provider.getConfigMap.get("kerberosEnable");
    kerberosEnable match {
      case flag: Boolean =>
        if (flag) {
          val principalName = provider.getConfigMap.get("principalName").asInstanceOf[String]
          val keytabPath = provider.getConfigMap.get("keytabPath").asInstanceOf[String]
          val krb5Path = provider.getConfigMap.get("krb5Path").asInstanceOf[String]
          val requestKuu = new RequestKerberosUrlUtils(principalName, keytabPath, krb5Path, false)
          val response =
            requestKuu.callRestUrl(rmWebAddress + "/ws/v1/cluster/" + url, principalName)
          httpResponse = response;
        } else {
          val response = YarnResourceRequester.httpClient.execute(httpGet)
          httpResponse = response
        }
      case _ =>
        val response = YarnResourceRequester.httpClient.execute(httpGet)
        httpResponse = response
    }
    parse(EntityUtils.toString(httpResponse.getEntity()))
  }

  def getAndUpdateActiveRmWebAddress(provider: ExternalResourceProvider): String = {
    // todo check if it will stuck for many requests
    val haAddress = provider.getConfigMap.get("rmWebAddress").asInstanceOf[String]
    var activeAddress = rmAddressMap.get(haAddress)
    if (StringUtils.isBlank(activeAddress)) haAddress.intern().synchronized {
      if (StringUtils.isBlank(activeAddress)) {
        if (logger.isDebugEnabled()) {
          logger.debug(
            s"Cannot find value of haAddress : ${haAddress} in cacheMap with size ${rmAddressMap.size()}"
          )
        }
        if (StringUtils.isNotBlank(haAddress)) {
          haAddress
            .split(RMConfiguration.DEFAULT_YARN_RM_WEB_ADDRESS_DELIMITER.getValue)
            .foreach(address => {
              Utils.tryCatch {
                val response = getResponseByUrl("info", address, provider)
                response \ "clusterInfo" \ "haState" match {
                  case state: JString =>
                    if (HASTATE_ACTIVE.equalsIgnoreCase(state.s)) {
                      activeAddress = address
                    } else {
                      logger.warn(s"Resourcemanager : ${address} haState : ${state.s}")
                    }
                  case _ =>
                }
              } { case e: Exception =>
                logger.error("Get Yarn resourcemanager info error, " + e.getMessage, e)
              }
            })
        }
        if (StringUtils.isNotBlank(activeAddress)) {
          if (logger.isDebugEnabled()) {
            logger.debug(s"Put (${haAddress}, ${activeAddress}) to cacheMap.")
          }
          rmAddressMap.put(haAddress, activeAddress)
        } else {
          throw new RMErrorException(
            GET_YARN_EXCEPTION.getErrorCode,
            MessageFormat.format(GET_YARN_EXCEPTION.getErrorDesc(), haAddress)
          )
        }
      }
    }
    if (logger.isDebugEnabled()) {
      logger.debug(s"Get active rm address : ${activeAddress} from haAddress : ${haAddress}")
    }
    activeAddress
  }

  override def reloadExternalResourceAddress(
      provider: ExternalResourceProvider
  ): java.lang.Boolean = {
    if (null != provider) {
      val rmWebHaAddress = provider.getConfigMap.get("rmWebAddress").asInstanceOf[String]
      rmAddressMap.remove(rmWebHaAddress)
      getAndUpdateActiveRmWebAddress(provider)
    }
    true
  }

}

object YarnResourceRequester extends Logging {

  private val httpClient = HttpClients.createDefault()

  def init(): Unit = {
    addShutdownHook()
  }

  def addShutdownHook(): Unit = {
    logger.info("Register shutdown hook to release httpclient connection")
    Utils.addShutdownHook(httpClient.close())
  }

}

File: linkis-computation-governance/linkis-manager/linkis-application-manager/src/main/java/org/apache/linkis/manager/rm/external/yarn/YarnResourceRequester.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.linkis.manager.rm.external.yarn;

import org.apache.linkis.manager.am.util.LinkisUtils;
import org.apache.linkis.manager.common.conf.RMConfiguration;
import org.apache.linkis.manager.common.entity.resource.*;
import org.apache.linkis.manager.common.exception.RMErrorException;
import org.apache.linkis.manager.common.exception.RMWarnException;
import org.apache.linkis.manager.rm.external.domain.ExternalAppInfo;
import org.apache.linkis.manager.rm.external.domain.ExternalResourceIdentifier;
import org.apache.linkis.manager.rm.external.domain.ExternalResourceProvider;
import org.apache.linkis.manager.rm.external.request.ExternalResourceRequester;
import org.apache.linkis.manager.rm.utils.RequestKerberosUrlUtils;

import org.apache.commons.lang3.StringUtils;
import org.apache.commons.lang3.tuple.Pair;
import org.apache.http.HttpHeaders;
import org.apache.http.HttpResponse;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClients;
import org.apache.http.util.EntityUtils;

import java.io.IOException;
import java.text.MessageFormat;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import static org.apache.linkis.manager.common.errorcode.ManagerCommonErrorCodeSummary.*;

public class YarnResourceRequester implements ExternalResourceRequester {
  private static final Logger logger = LoggerFactory.getLogger(YarnResourceRequester.class);

  private final String HASTATE_ACTIVE = "active";
  private static final ObjectMapper objectMapper = new ObjectMapper();
  private final Map<String, String> rmAddressMap = new ConcurrentHashMap<>();

  private String getAuthorizationStr(ExternalResourceProvider provider) {
    String user = (String) provider.getConfigMap().getOrDefault("user", "");
    String pwd = (String) provider.getConfigMap().getOrDefault("pwd", "");
    String authKey = user + ":" + pwd;
    return Base64.getMimeEncoder().encodeToString(authKey.getBytes());
  }

  @Override
  public NodeResource requestResourceInfo(
      ExternalResourceIdentifier identifier, ExternalResourceProvider provider) {
    String rmWebAddress = getAndUpdateActiveRmWebAddress(provider);
    logger.info("rmWebAddress: " + rmWebAddress);

    String queueName = ((YarnResourceIdentifier) identifier).getQueueName();
    String realQueueName = "root." + queueName;

    return LinkisUtils.tryCatch(
        () -> {
          Pair<YarnResource, YarnResource> yarnResource =
              getResources(rmWebAddress, realQueueName, queueName, provider);

          CommonNodeResource nodeResource = new CommonNodeResource();
          nodeResource.setMaxResource(yarnResource.getKey());
          nodeResource.setUsedResource(yarnResource.getValue());
          return nodeResource;
        },
        t -> {
          throw new RMErrorException(
              YARN_QUEUE_EXCEPTION.getErrorCode(), YARN_QUEUE_EXCEPTION.getErrorDesc(), t);
        });
  }

  public Optional<YarnResource> maxEffectiveHandle(
      Optional<JsonNode> queueValue,
      String rmWebAddress,
      String queueName,
      ExternalResourceProvider provider) {
    try {
      JsonNode metrics = getResponseByUrl("metrics", rmWebAddress, provider);
      JsonNode clusterMetrics = metrics.path("clusterMetrics");
      long totalMemory = clusterMetrics.path("totalMB").asLong();
      long totalCores = clusterMetrics.path("totalVirtualCores").asLong();
      if (queueValue.isPresent()) {
        JsonNode jsonNode = queueValue.get();
        double absoluteCapacity = jsonNode.path("absoluteCapacity").asDouble();

        YarnResource yarnResource =
            new YarnResource(
                (long) Math.floor(absoluteCapacity * totalMemory * 1024L * 1024L / 100),
                (int) Math.floor(absoluteCapacity * totalCores / 100),
                0,
                queueName,
                "");
        return Optional.of(yarnResource);
      }
      return Optional.empty();
    } catch (Exception e) {
      logger.warn("maxEffectiveHandle parse failed", e);
      return Optional.empty();
    }
  }

  public static Optional getQueue(JsonNode queues, String realQueueName) {
    if (queues instanceof ArrayNode) {
      for (JsonNode q : (ArrayNode) queues) {
        String yarnQueueName = q.get("queueName").asText();
        if (yarnQueueName.equals(realQueueName)) {
          return Optional.of(q);
        } else if (realQueueName.startsWith(yarnQueueName + ".")) {
          Optional childQueue = getQueue(getChildQueues(q), realQueueName);
          if (childQueue.isPresent()) {
            return childQueue;
          }
        }
      }
      return Optional.empty();
    } else if (queues instanceof ObjectNode) {
      ObjectNode queueObj = (ObjectNode) queues;
      JsonNode queueName = queueObj.get("queueName");
      if (queueName != null && queueName.asText().equals(realQueueName)) {
        return Optional.of(queueObj);
      } else {
        JsonNode childQueues = queueObj.get("childQueues");
        if (childQueues == null) {
          return Optional.empty();
        } else {
          return getQueue(childQueues, realQueueName);
        }
      }
    } else {
      return Optional.empty();
    }
  }

  public static JsonNode getChildQueues(JsonNode resp) {
    JsonNode queues = resp.get("childQueues").get("queue");
    if (queues != null
        && !queues.isNull()
        && !queues.isMissingNode()
        && queues.isArray()
        && ((ArrayNode) queues).size() > 0) {
      return queues;
    } else {
      return resp.get("childQueues");
    }
  }

  public Optional<JsonNode> getQueueOfCapacity(JsonNode queues, String realQueueName) {
    if (queues.isArray()) {
      for (JsonNode q : queues) {
        String yarnQueueName = q.get("queueName").asText();
        if (yarnQueueName.equals(realQueueName)) {
          return Optional.of(q);
        } else if (q.has("queues")) {
          Optional<JsonNode> matchQueue = getQueueOfCapacity(q.get("queues"), realQueueName);
          if (matchQueue.isPresent()) {
            return matchQueue;
          }
        }
      }
      return Optional.empty();
    } else if (queues.isObject()) {
      if (queues.has("queueName") && queues.get("queueName").asText().equals(realQueueName)) {
        return Optional.of(queues);
      } else if (queues.has("queues")) {
        Optional<JsonNode> matchQueue = getQueueOfCapacity(queues.get("queues"), realQueueName);
        if (matchQueue.isPresent()) {
          return matchQueue;
        }
      }
      return Optional.empty();
    } else {
      return Optional.empty();
    }
  }

  static JsonNode getChildQueuesOfCapacity(JsonNode resp) {
    // = resp \ "queues" \ "queue"
    return resp.path("queues").path("queue");
  }

  public Pair<YarnResource, YarnResource> getResources(
      String rmWebAddress,
      String realQueueName,
      String queueName,
      ExternalResourceProvider provider) {
    JsonNode resp = getResponseByUrl("scheduler", rmWebAddress, provider);
    JsonNode schedulerInfo = resp.path("scheduler").path("schedulerInfo");
    String schedulerType = schedulerInfo.path("type").asText();
    if ("capacityScheduler".equals(schedulerType)) {
      realQueueName = queueName;
      JsonNode childQueues = getChildQueuesOfCapacity(schedulerInfo);
      Optional<JsonNode> queue = getQueueOfCapacity(childQueues, realQueueName);
      if (!queue.isPresent()) {
        logger.debug(
            "cannot find any information about queue " + queueName + ", response: " + resp);
        throw new RMWarnException(
            YARN_NOT_EXISTS_QUEUE.getErrorCode(),
            MessageFormat.format(YARN_NOT_EXISTS_QUEUE.getErrorDesc(), queueName));
      }
      return Pair.of(
          maxEffectiveHandle(queue, rmWebAddress, queueName, provider).get(),
          getYarnResource(queue.map(node -> node.path("resourcesUsed")), queueName).get());

    } else if ("fairScheduler".equals(schedulerType)) {
      JsonNode childQueues = getChildQueues(schedulerInfo.path("rootQueue"));
      Optional<JsonNode> queue = getQueue(childQueues, realQueueName);
      if (!queue.isPresent()) {
        logger.debug(
            "cannot find any information about queue " + queueName + ", response: " + resp);
        throw new RMWarnException(
            YARN_NOT_EXISTS_QUEUE.getErrorCode(),
            MessageFormat.format(YARN_NOT_EXISTS_QUEUE.getErrorDesc(), queueName));
      }
      Optional<JsonNode> maxResources = queue.map(node -> node.path("maxResources"));
      Optional<JsonNode> usedResources = queue.map(node -> node.path("usedResources"));
      return Pair.of(
          getYarnResource(maxResources, queueName).get(),
          getYarnResource(usedResources, queueName).get());

    } else {
      logger.debug(
          "only support fairScheduler or capacityScheduler, schedulerType: "
              + schedulerType
              + ", response: "
              + resp);
      throw new RMWarnException(
          ONLY_SUPPORT_FAIRORCAPA.getErrorCode(),
          MessageFormat.format(ONLY_SUPPORT_FAIRORCAPA.getErrorDesc(), schedulerType));
    }
  }

  public static Optional<YarnResource> getYarnResource(
      Optional<JsonNode> jsonNode, String queueName) {
    if (jsonNode.isPresent()) {
      JsonNode r = jsonNode.get();
      return Optional.of(
          new YarnResource(
              r.get("memory").asLong() * 1024L * 1024L, r.get("vCores").asInt(), 0, queueName, ""));
    }
    return Optional.empty();
  }

  public static Optional<YarnResource> getAllocatedYarnResource(
      Optional<JsonNode> jsonNode, String queueName) {
    if (jsonNode.isPresent()) {
      JsonNode r = jsonNode.get();
      return Optional.of(
          new YarnResource(
              r.get("allocatedMB").asLong() * 1024L * 1024L,
              r.get("allocatedVCores").asInt(),
              0,
              queueName,
              ""));
    }
    return Optional.empty();
  }

  @Override
  public List<ExternalAppInfo> requestAppInfo(
      ExternalResourceIdentifier identifier, ExternalResourceProvider provider) {

    String rmWebAddress = getAndUpdateActiveRmWebAddress(provider);

    String queueName = ((YarnResourceIdentifier) identifier).getQueueName();
    String realQueueName = "root." + queueName;

    JsonNode resp = getResponseByUrl("apps", rmWebAddress, provider).path("apps").path("app");
    if (resp.isMissingNode()) {
      return new ArrayList<>();
    }
    ArrayNode appArray = (ArrayNode) resp;

    List<ExternalAppInfo> appInfoList = new ArrayList<>();
    Iterator<JsonNode> iterator = appArray.elements();
    while (iterator.hasNext()) {
      JsonNode app = iterator.next();
      String yarnQueueName = app.get("queue").asText();
      String state = app.get("state").asText();
      if (yarnQueueName.equals(realQueueName)
          && (state.equals("RUNNING") || state.equals("ACCEPTED"))) {
        String id = app.get("id").asText();
        String user = app.get("user").asText();
        String applicationType = app.get("applicationType").asText();
        Optional<YarnResource> yarnResource =
            getAllocatedYarnResource(Optional.ofNullable(app), queueName);
        if (yarnResource.isPresent()) {
          YarnAppInfo appInfo =
              new YarnAppInfo(id, user, state, applicationType, yarnResource.get());
          appInfoList.add(appInfo);
        }
      }
    }
    return appInfoList;
  }

  @Override
  public ResourceType getResourceType() {
    return ResourceType.Yarn;
  }

  private JsonNode getResponseByUrl(
      String url, String rmWebAddress, ExternalResourceProvider provider) {

    HttpGet httpGet = new HttpGet(rmWebAddress + "/ws/v1/cluster/" + url);
    httpGet.addHeader("Accept", "application/json");
    Object authorEnable = provider.getConfigMap().get("authorEnable");
    HttpResponse httpResponse = null;
    if (authorEnable instanceof Boolean) {
      if ((Boolean) authorEnable) {
        httpGet.addHeader(HttpHeaders.AUTHORIZATION, "Basic " + getAuthorizationStr(provider));
      }
    }
    Object kerberosEnable = provider.getConfigMap().get("kerberosEnable");
    if (kerberosEnable instanceof Boolean) {
      if ((Boolean) kerberosEnable) {
        String principalName = (String) provider.getConfigMap().get("principalName");
        String keytabPath = (String) provider.getConfigMap().get("keytabPath");
        String krb5Path = (String) provider.getConfigMap().get("krb5Path");
        if (StringUtils.isNotBlank(krb5Path)) {
          logger.warn(
              "krb5Path: {} has been specified, but not allow to be set to avoid conflict",
              krb5Path);
        }
        RequestKerberosUrlUtils requestKuu =
            new RequestKerberosUrlUtils(principalName, keytabPath, false);
        HttpResponse response =
            requestKuu.callRestUrl(rmWebAddress + "/ws/v1/cluster/" + url, principalName);
        httpResponse = response;
      } else {
        HttpResponse response = null;
        try {
          CloseableHttpClient httpClient = HttpClients.createDefault();
          response = httpClient.execute(httpGet);
        } catch (IOException e) {
          logger.warn("getResponseByUrl failed", e);
          throw new RMErrorException(
              YARN_QUEUE_EXCEPTION.getErrorCode(), YARN_QUEUE_EXCEPTION.getErrorDesc(), e);
        }
        httpResponse = response;
      }
    } else {
      HttpResponse response = null;
      try {
        CloseableHttpClient httpClient = HttpClients.createDefault();
        response = httpClient.execute(httpGet);
      } catch (IOException e) {
        logger.warn("getResponseByUrl failed", e);
        throw new RMErrorException(
            YARN_QUEUE_EXCEPTION.getErrorCode(), YARN_QUEUE_EXCEPTION.getErrorDesc(), e);
      }
      httpResponse = response;
    }

    String entityString = "";
    try {
      entityString = EntityUtils.toString(httpResponse.getEntity());
    } catch (IOException e) {
      logger.warn("getResponseByUrl failed", e);
      throw new RMErrorException(
          YARN_QUEUE_EXCEPTION.getErrorCode(), YARN_QUEUE_EXCEPTION.getErrorDesc(), e);
    }
    JsonNode jsonNode = null;
    try {
      jsonNode = objectMapper.readTree(entityString);
    } catch (Exception e) {
      logger.warn("getResponseByUrl failed", e);
      throw new RMErrorException(
          YARN_QUEUE_EXCEPTION.getErrorCode(), YARN_QUEUE_EXCEPTION.getErrorDesc(), e);
    }

    return jsonNode;
  }

  public String getAndUpdateActiveRmWebAddress(ExternalResourceProvider provider) {
    // todo check if it will stuck for many requests
    String haAddress = (String) provider.getConfigMap().get("rmWebAddress");
    String activeAddress = rmAddressMap.get(haAddress);
    if (StringUtils.isBlank(activeAddress)) {
      synchronized (haAddress.intern()) {
        if (StringUtils.isBlank(activeAddress)) {
          if (logger.isDebugEnabled()) {
            logger.debug(
                "Cannot find value of haAddress : "
                    + haAddress
                    + " in cacheMap with size "
                    + rmAddressMap.size());
          }
          if (StringUtils.isNotBlank(haAddress)) {
            String[] addresses =
                haAddress.split(RMConfiguration.DEFAULT_YARN_RM_WEB_ADDRESS_DELIMITER.getValue());
            for (String address : addresses) {
              try {
                JsonNode response = getResponseByUrl("info", address, provider);
                JsonNode haStateValue = response.path("clusterInfo").path("haState");
                if (!haStateValue.isMissingNode() && haStateValue.isTextual()) {
                  String haState = haStateValue.asText();
                  if (HASTATE_ACTIVE.equalsIgnoreCase(haState)) {
                    activeAddress = address;
                  } else {
                    logger.warn("Resourcemanager : " + address + " haState : " + haState);
                  }
                }
              } catch (Exception e) {
                logger.error("Get Yarn resourcemanager info error, " + e.getMessage(), e);
              }
            }
          }
          if (StringUtils.isNotBlank(activeAddress)) {
            if (logger.isDebugEnabled()) {
              logger.debug("Put (" + haAddress + ", " + activeAddress + ") to cacheMap.");
            }
            rmAddressMap.put(haAddress, activeAddress);
          } else {
            throw new RMErrorException(
                GET_YARN_EXCEPTION.getErrorCode(),
                MessageFormat.format(GET_YARN_EXCEPTION.getErrorDesc(), haAddress));
          }
        }
      }
    }
    if (logger.isDebugEnabled()) {
      logger.debug("Get active rm address : " + activeAddress + " from haAddress : " + haAddress);
    }
    return activeAddress;
  }

  @Override
  public Boolean reloadExternalResourceAddress(ExternalResourceProvider provider) {
    if (null != provider) {
      String rmWebHaAddress = (String) provider.getConfigMap().get("rmWebAddress");
      rmAddressMap.remove(rmWebHaAddress);
      getAndUpdateActiveRmWebAddress(provider);
    }
    return true;
  }
}
